worker/src/enclave/api.rs

```diff
@@ -183,214 +110,61 @@ pub fn enclave_init() -> SgxResult<SgxEnclave> {
         }
     }
 
-    let mut status = sgx_status_t::SGX_SUCCESS;
-    // call the enclave's init fn
-    let result = unsafe { init(enclave.geteid(), &mut status) };
-    if status != sgx_status_t::SGX_SUCCESS {
-        return Err(status);
-    }
-    if result != sgx_status_t::SGX_SUCCESS {
-        return Err(result);
-    }
-    Ok(enclave)
+    // create an enclave API and initialize it
+    let enclave_api = Enclave::new(enclave);
+    enclave_api.init()?;
+
+    Ok(enclave_api)
 }
 
-pub fn enclave_init_chain_relay(
-    eid: sgx_enclave_id_t,
+pub fn enclave_init_chain_relay<E: EnclaveApi>(
     genesis_header: Header,
     authority_list: VersionedAuthorityList,
     authority_proof: Vec<Vec<u8>>,
-) -> SgxResult<Header> {
```

Comment:
The functions in this source file originally were like an API wrapper for the e-calls. Now they all use the `EnclaveApi` trait. The reason why they still exist, is because they do some additional work for encoding&decoding arguments, deserialization and logging. We should discuss if that functionality should stay here, or if we move that to the `enclave-api` as well, making it a little more heavy weight than originally intended.

Reply:
Done 👍 Header is a type parameter, and `VersionedAuthorityList` is taken as `Vec<u8>` (to avoid importing `sp_grandpa_finality` as dependency)

Reply:
Looks very nice now!

Regarding the `VersionedAuthority` though, I personally prefer to type the interfaces as strong as possible, but it does not matter much in this case.

Reply:
I we want to be super-rustacean-strongly-typed-nerds, which I (maybe shamefully) like to be, we could even use the `sp_runtime::traits::Header` as generic parameter for the header. To make sure that only something that implements indeed represends a header is returned.

Reply:
It's anyway strongly typed, by definition in Rust? But the question is how specific should an input parameter be in this case. Unintentional mix-ups should be prevented, I agree. In this case I think it's are trade-off between being less specific, but avoiding some dependencies. Or be more specific and having to depend on `sp_runtime` ?

Reply:
I'd be more specific and don't care about the dependency in favor of disambiguation.

As the calling create is 99% probable to anyhow depend on `sp_runtime`, there should be no overhead.

Reply:
done 👍 added `sp-finality-grandpa` and `sp-runtime` to the dependencies, and bound the type parameter by `sp_runtime::Header` and made the authority list a `VersionedAuthorityList` again.